snuba/datasets/cdc/cdcprocessors.py

```diff
@@ -0,0 +1,56 @@
+from snuba import processor
+from snuba.processor import MessageProcessor
+
+
+class CdcProcessor(MessageProcessor):
+
+    def __init__(self, pg_table):
+        self.pg_table = pg_table
+
+    def _process_begin(self, xid):
+        pass
+
+    def _process_commit(self):
+        pass
+
+    def _process_insert(self, xid, columnnames, columnvalues):
+        pass
+
+    def _process_update(self, xid, key, columnnames, columnvalues):
+        pass
+
+    def _process_delete(self, xid, key):
+        pass
+
+    def process_message(self, value):
+        assert isinstance(value, dict)
+        event = value['event']
+        if event == 'begin':
+            xid = value['xid']
+            message = self._process_begin(xid)
+        elif event == 'commit':
+            message = self._process_commit()
+        elif event == 'change':
+            table_name = value['table']
+            if table_name != self.pg_table:
+                return None
+
+            operation = value['kind']
+            xid = value['xid']
+            if operation == 'insert':
+                message = self._process_insert(
+                    xid, value['columnnames'], value['columnvalues'])
+            elif operation == 'update':
+                message = self._process_update(
+                    xid, value['oldkeys'], value['columnnames'], value['columnvalues'])
```

Comment:
It seems like there's an implicit assumption that the replica identify for a row is not changing. We should make that explicit with an assertion or something.

Reply:
Actually reading this again. I don't think there is such an assumption:
1) if the *values* of the identifier change, then you find the new values in the value["columnvalues"] field.
2) if we are changing the structure of the identifier (whether it is even possible) we just don't have that information from the replication log. So there is nothing we can assert on.
Or am I interpreting your comment the wrong way ?

Reply:
I need to reread the wal2json implementation to be sure, but I swore these ended up in a separate data structure if the identity is mutated. (Specifically [here](https://github.com/getsentry/wal2json/blob/5a453aa8331409889c44e991a3bf2ba050ae59ab/wal2json.c#L471-L476), but this might not be reachable from the path I'm thinking of. I haven't walked it back yet, so I may very well be wrong.)

Reply:
So this is the output when I change the primary key of a record in sentry_groupedmessage:

```
{"event":"begin","xid":2437538}
{"event":"change",
  "xid":2437538,
   "kind":"update",
    "schema":"public",
    "table":"sentry_groupedmessage",
    "columnnames":[
            "id","logger","level","message","view","status","times_seen","last_seen",
            "first_seen","data","score","project_id","time_spent_total","time_spent_count",
           "resolved_at","active_at","is_public","platform","num_comments",
           "first_release_id","short_id"],
     "columntypes":[
            "bigint","character varying(64)","integer","text","charactervarying(200)",
             "integer","integer","timestamp with time zone","timestamp with time zone",
             "text","integer","bigint","integer","integer","timestamp with time zone",
              "timestamp with time zone","boolean","character varying(64)",
               "integer","bigint","bigint"],
       "columnvalues":[
               51,"",40,
               "get KeyError None sentry/eventtypes/manager.py /api/0/projects/{organization_slug}/{project_slug}/events/{event_id}/",
               "bla",0,1,"2019-06-16 06:21:39+00","2019-06-16 06:21:39+00",
               "eJw1TrsOgzAM3PMV3WACwiNturdLpS5IXSsLDEIKIUoCEn9fG9EsOd",
                1560666100,1,0,0,null,"2019-06-16 06:21:39+00",false,"python",0,9,132],
       "oldkeys":{
             "keynames":["id"],"keytypes":["bigint"],"keyvalues":[50]}}
{"event":"commit"}
```

This has the same structure I find when I make any other sql UPDATE.
From the wal2json.c code it seems to me that, depending whether we are changing the key or not, wal2json will either pick the old key or the new one, which in both case ends up being the key of the object to change and the data structure does not change.

Actually, what I will do is to forbid, for now with an assertion, changes of the primary key because that would mess us up with clickhouse since, we would fail merging the records.

Reply:
I was misreading this earlier — I thought the column sets and the key sets were disjoint. This makes sense now.

Reply:
yeah, I think wal2json output can be confusing. We may consider changing it since we basically forked the project